leakpro/attacks/mia_attacks/lira.py

```diff
@@ -0,0 +1,222 @@
+        # print(np.count_nonzero(mask))
+        # print(np.count_nonzero(~mask))
+
+        # mask = ShadowModelHandler().get_in_indices_mask(self.num_shadow_models, self.audit_dataset["data"][self.audit_dataset["in_members"]])
+        # print(np.count_nonzero(mask))
+        # print(np.count_nonzero(~mask))
+        
+        self.in_indices_mask = ShadowModelHandler().get_in_indices_mask(self.num_shadow_models, self.audit_dataset["data"])
+        
+        self.audit_data = self.population.subset(self.audit_dataset["data"])
+        
+        # # Extract a matching subset from the population dataset
+        
+        # Calculate logits for all shadow models
+        print(f"Calculating the logits for all {self.num_shadow_models} shadow models")
+        self.shadow_models_logits = np.swapaxes(np.array(self.signal(self.shadow_models, self.audit_data)), 0, 1)
+        
+        # Calculate logits for the target model
+        print(f"Calculating the logits for the target model")
+        self.target_logits = np.array(self.signal([self.target_model], self.audit_data)).squeeze()     
+
+    def run_attack(self):
+        """
+        Runs the attack on the target model and dataset, computing and returning
+        the result(s) of the metric, which assess privacy risks or data leakage.
+    
+        This method evaluates how the target model's output (logits) for a specific dataset 
+        compares to the output of shadow models to determine if the dataset was part of the 
+        model's training data or not.
+    
+        Returns:
+            Result(s) of the metric. An object containing the metric results, including predictions,
+            true labels, and signal values.
+        """
+        
+        score = []  # List to hold the computed probability scores for each sample
+        
+        # If global standard deviation is to be used, calculate it from all logits of shadow models
+        if self.fixed_variance:
+            # in_std = np.nanstd(self.shadow_models_logits[self.in_indices_mask].flatten())
+            out_std = np.nanstd(self.shadow_models_logits[~self.in_indices_mask].flatten())
+            
+        # Iterate over each sample in the audit dataset with a progress bar
+        for i, mask in tqdm(enumerate(self.in_indices_mask)):
+                
+            # Extract logits from shadow models for the current sample
+            shadow_models_logits = self.shadow_models_logits[i, :]
+            out_mean = np.nanmean(shadow_models_logits[~mask])  # Calculate the mean of shadow model logits
+            
+            # Get the logit from the target model for the current sample
+            target_logit = self.target_logits[i]
+    
+            # Calculate the log probability density function value
+            if not self.fixed_variance:
+                out_std = np.nanstd(shadow_models_logits[~mask])
+                # in_std = np.nanstd(shadow_models_logits[mask])
```

Comment:
in general, I think we should avoid using np.nanstd and np.nanmean. We dont get to see if there are errors because the nans are not visible to us.

Reply:
Yes, for sure. Will remove them